plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,127 @@
+import io.trino.testing.AbstractTestQueryFramework;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Collections.synchronizedMap;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Test(singleThreaded = true) // this class is stateful, see fields
+public abstract class BaseJdbcConnectionCreationTest
+        extends AbstractTestQueryFramework
+{
+    protected ConnectionCountingConnectionFactory connectionFactory;
+
+    @BeforeClass
+    public void verifySetup()
+    {
+        // Test expects connectionFactory to be provided with AbstractTestQueryFramework.createQueryRunner implementation
+        requireNonNull(connectionFactory, "connectionFactory is null");
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected void assertJdbcConnections(@Language("SQL") String query, int expectedJdbcConnectionsCount, Optional<String> errorMessage)
+    {
+        int before = connectionFactory.openConnections.get();
+        if (errorMessage.isPresent()) {
+            assertQueryFails(query, errorMessage.get());
+        }
+        else {
+            getQueryRunner().execute(query);
+        }
+        int after = connectionFactory.openConnections.get();
+        assertThat(after - before).isEqualTo(expectedJdbcConnectionsCount);
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected static class ConnectionCountingConnectionFactory
+            implements ConnectionFactory
+    {
+        // Map from connection to a fake exception (holds stacktrace) pointing to the place where the connection was created
+        private final Map<Connection, Exception> connectionCreations = synchronizedMap(new IdentityHashMap<>());
+        private final AtomicInteger openConnections = new AtomicInteger();
+        private final ConnectionFactory delegate;
+
+        public ConnectionCountingConnectionFactory(DriverConnectionFactory delegate)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        public Connection openConnection(ConnectorSession session)
+                throws SQLException
+        {
+            openConnections.incrementAndGet();
+            Connection connection = delegate.openConnection(session);
+            Exception previous = connectionCreations.put(connection, new Exception("STACKTRACE"));
+            if (previous != null) {
+                IllegalStateException exception = new IllegalStateException("Same connection %s returned twice from delegate".formatted(connection));
```

Comment:
why is this a problem? It's unclear to me.

Reply:
`connectionCreations` do not support two connections at a time at the moment.